### PR TITLE
HIVE-26911: Renaming a translated external table with a specified loc…

### PR DESCRIPTION
**Pull request by saihemanth-cloudera**
…ation fails with location already exists

### What changes were proposed in this pull request?
Fixes a issue where Renaming a translated external table with a specified location fails with location already exists exception.

### Why are the changes needed?
Currently we cannot rename a translated external table

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
unit tests

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java [hunk 1]
```diff
@@ -270,8 +270,12 @@ public void alterTable(RawStore msdb, Warehouse wh, String catName, String dbnam
 
             // get new location
             assert(isReplicated == HMSHandler.isDbReplicationTarget(db));
-            if (renamedTranslatedToExternalTable) {
+            if (renamedTranslatedToExternalTable && !tableInSpecifiedLoc) {
               destPath = new Path(newt.getSd().getLocation());
```
Comment (dengzhhu653) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java:274]: What happen if the path for the old table `b` is `file:${system:test.tmp.dir}/tmp.db/b`?  Does it throw the same exception?

Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java:274]: nit: we can build the `destPath` under `if (renamedTranslatedToExternalTablc)`:
```java
if (renamedTranslatedToExternalTablc) {
   if (tableInSpecifiedLoc) {
     ....
   } else {
    ...
   }
}

Comment (saihemanth-cloudera) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java:274]: @dengzhhu653 - if you explicitly specify a table location for old table b: as `file:${system:test.tmp.dir}/tmp.db/b` then it fails with the same exception. Only if you don't specify a location the query would pass

File: ql/src/test/queries/clientpositive/rename_translated_external_table.q [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+set hive.fetch.task.conversion=none;
+set hive.compute.query.using.stats=false;
+
+create database if not exists tmp;
+
+use tmp;
+
+create table b(s string) stored as parquet location 'file:${system:test.tmp.dir}/tmp.db/some_location';
+
+desc formatted b;
```
Comment (dengzhhu653) [inline ql/src/test/queries/clientpositive/rename_translated_external_table.q:11]: Output of the query does not show the details about the location, so I think it better to do like this:
```sql
insert into table b values('a'),('b'),('c');
select * from b;
```
after renaming b to bb, then
```
select * from bb;
-- new table and location should be available after rename
create table b(s string) stored as parquet location 'file:${system:test.tmp.dir}/tmp.db/some_location';
select * from b;  -- the result should be empty
```

File: ql/src/test/queries/clientpositive/rename_translated_external_table.q [hunk 3]
```diff
@@ -0,0 +1,18 @@
+set metastore.metadata.transformer.class=org.apache.hadoop.hive.metastore.MetastoreDefaultTransformer;
+set hive.fetch.task.conversion=none;
+set hive.compute.query.using.stats=false;
```
Comment (dengzhhu653) [inline ql/src/test/queries/clientpositive/rename_translated_external_table.q:3]: These properties seem no use for this case?

Comment (saihemanth-cloudera) [reply ql/src/test/queries/clientpositive/rename_translated_external_table.q:3]: ack

File: ql/src/test/queries/clientpositive/rename_translated_external_table.q (hunk 2 above)
Comment (saihemanth-cloudera) [reply ql/src/test/queries/clientpositive/rename_translated_external_table.q:11]: ack

**Comment by dengzhhu653** [comment]
Change looks good to me, and please fix the failed test.

**Review by dengzhhu653 — APPROVED** [review]

**Pull request merged**